### PR TITLE
Created an extension mechanism to auto-fold Text/Diagram

### PR DESCRIPTION
**Pull request by mohamed-abead**
## Summary
Users have to scroll a lot of lines of code for Text and Diagram that are often irrelevant to what users are trying to work on. Thus, Text and Diagram lines tend to clutter developers screens, and are often too long (spanning hundreds or thousands of lines of code.) We tackle that by automatically folding Text/Diagram elements as soon as users enter Text mode. Moreover, this PR creates an extension mechanism to allow for other elements to be auto-folded in the future, if needed. Finally, this PR adds a button to the top right of the screen to allow users to fold/unfold those auto-folding elements easily. 

## How did you test this change?
- [ ] Test(s) added
- [x] Manual testing (please provide screenshots/recordings)
- [ ] No testing (please provide an explanation)

I manually tested:
1 - Having a single Diagram, or a single Text element
2 - Having multiple Diagram and Text elements
3 - Typing in the editor in different places, or right before or after some of the folded elements
  * Also, tried manually unfolding and then typing somewhere, and then folding again using the button or manually

4 - Switching back and forth between Text mode and form mode
  * Also tried adding new Text and/or Diagram elements and switching back and forth as I did so

5 - Tested having empty Text and Diagram elements
6 - Tested folding/unfolding elements other than Text and Diagram, and making sure they don't affect the behavior of the button, or the auto-folding action
7 - Manually unfolding Text/Diagram elements and switching back to Form mode and back to Text mode again

[Demo Video](https://github.com/finos/legend-studio/assets/73408381/1f066684-15a2-4fbf-af51-e1190d09e71c)

### CODE REVIEW

**Comment by YannanGao-gs** [comment]
demo video:

https://github.com/finos/legend-studio/assets/73408381/1f066684-15a2-4fbf-af51-e1190d09e71c

**Comment by YannanGao-gs** [comment]
link:https://github.com/finos/legend-studio/assets/73408381/1f066684-15a2-4fbf-af51-e1190d09e71c

File: packages/legend-application-studio/src/stores/LegendStudioApplicationPlugin.ts [hunk 1]
```diff
@@ -321,4 +321,10 @@ export interface DSL_LegendStudioApplicationPlugin_Extension
    * (e.g. Class, Enum in ###Pure)
    */
   getExtraPureGrammarParserElementSnippetSuggestionsGetters?(): PureGrammarParserElementSnippetSuggestionsGetter[];
+
+  /**
+   * Get a string of the Pure grammar element name for auto-folding the element
+   * (e.g. Diagram, or Text)
+   */
+  getAutoFoldingCandidateToken?(): string[];
```
Comment (akphi) [inline packages/legend-application-studio/src/stores/LegendStudioApplicationPlugin.ts:329]: @YannanGao-gs a small `TODO`

```suggestion
  getExtraGrammarTextEditorAutoFoldingElementCreatorKeywords?(): string[];
```

**Review by akphi — APPROVED** [review]

**Comment by akphi** [comment]
@mohamed-abead Thank you so much!

File: packages/legend-application-studio/src/stores/LegendStudioApplicationPlugin.ts (hunk 1 above)
Comment (YannanGao-gs) [reply packages/legend-application-studio/src/stores/LegendStudioApplicationPlugin.ts:329]: resolved by https://github.com/finos/legend-studio/pull/2599

**Pull request merged**